src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java

```diff
@@ -0,0 +1,173 @@
+package sparksoniq.jsoniq.runtime.iterator.operational;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function;
+import org.rumbledb.api.Item;
+import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.exceptions.TreatException;
+import sparksoniq.exceptions.UnexpectedTypeException;
+import sparksoniq.jsoniq.runtime.iterator.HybridRuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.functions.sequences.general.TreatAsClosure;
+import sparksoniq.jsoniq.runtime.iterator.functions.sequences.general.TypePromotionClosure;
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.semantics.types.ItemType;
+import sparksoniq.semantics.types.ItemTypes;
+import sparksoniq.semantics.types.SequenceType;
+
+import java.util.Collections;
+
+public class TypePromotionIterator extends HybridRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+    private final String _exceptionMessage;
+    private RuntimeIterator _iterator;
+    private SequenceType _sequenceType;
+
+    private ItemType itemType;
+    private String sequenceTypeName;
+
+    private Item _nextResult;
+    private int _childIndex;
+
+    public TypePromotionIterator(
+            String exceptionMessage,
+            RuntimeIterator iterator,
+            SequenceType sequenceType,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(Collections.singletonList(iterator), iteratorMetadata);
+        this._exceptionMessage = exceptionMessage;
+        this._iterator = iterator;
+        this._sequenceType = sequenceType;
+        itemType = _sequenceType.getItemType();
+        sequenceTypeName = ItemTypes.getItemTypeName(itemType.getType().toString());
+    }
+
+    @Override
+    public boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    public void resetLocal(DynamicContext context) {
+        _iterator.reset(_currentDynamicContext);
+        setNextResult();
+    }
+
+    @Override
+    public void closeLocal() {
+        _iterator.close();
+    }
+
+    @Override
+    public void openLocal() {
+        this._childIndex = 0;
+        _iterator.open(_currentDynamicContext);
+        this.setNextResult();
+    }
+
+
+    @Override
+    public Item nextLocal() {
+        if (this._hasNext) {
+            Item _currentResult = _nextResult;
+            setNextResult();
+            return _currentResult;
+        } else
+            throw new IteratorFlowException(RuntimeIterator.FLOW_EXCEPTION_MESSAGE, getMetadata());
+    }
+
+    private void setNextResult() {
+        _nextResult = null;
+        if (_iterator.hasNext()) {
+            _nextResult = _iterator.next();
+            if (_nextResult != null)
```

Comment:
next() never returns null if hasNext() was true, so this condition should always be true.

Reply:
That is right, thank you for spotting this